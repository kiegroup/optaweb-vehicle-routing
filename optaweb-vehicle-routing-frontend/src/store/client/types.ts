import { Viewport as LeafletViewport } from 'react-leaflet';
import { Action } from 'redux';

export enum ActionType {
  UPDATE_VIEWPORT = 'UPDATE_VIEWPORT',
  RESET_VIEWPORT = 'RESET_VIEWPORT',
}

export interface UpdateViewportAction extends Action<ActionType.UPDATE_VIEWPORT> {
  value: LeafletViewport;
}

export interface ResetViewportAction extends Action<ActionType.RESET_VIEWPORT> {
}

export interface UserViewport {
  isDirty: boolean;
  center: [number, number];
  zoom: number;
}

export type ViewportAction =
  | UpdateViewportAction
  | ResetViewportAction;
