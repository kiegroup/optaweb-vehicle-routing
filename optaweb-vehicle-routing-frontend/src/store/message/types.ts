import { Action } from 'redux';

export enum ActionType {
  RECEIVE_MESSAGE = 'RECEIVE_MESSAGE',
  READ_MESSAGE = 'READ_MESSAGE',
}

export interface ReceiveMessageAction extends Action<ActionType.RECEIVE_MESSAGE> {
  readonly payload: MessagePayload;
}

export interface ReadMessageAction extends Action<ActionType.READ_MESSAGE> {
  readonly id: string;
}

export type MessageAction = ReceiveMessageAction | ReadMessageAction;

export interface MessagePayload {
  readonly id: string;
  readonly text: string;
}

export interface Message extends MessagePayload {
  readonly status: 'new' | 'read';
}
