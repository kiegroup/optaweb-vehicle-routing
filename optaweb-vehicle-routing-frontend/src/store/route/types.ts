/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Action } from 'redux';

export interface LatLng {
  readonly lat: number;
  readonly lng: number;
}

export interface LatLngWithDescription extends LatLng {
  description: string;
}

export interface Location extends LatLng {
  readonly id: number;
  // TODO decide between optional, nullable and more complex structure (displayName, fullDescription, address, ...)
  readonly description?: string;
}

export interface Vehicle {
  readonly id: number;
  readonly name: string;
  readonly capacity: number;
}

export interface Route {
  readonly vehicle: Vehicle; // TODO change to vehicleId
  readonly visits: Location[];
}

export type LatLngTuple = [number, number];

export interface RouteWithTrack extends Route {
  readonly track: LatLngTuple[];
}

export interface RoutingPlan {
  readonly distance: string;
  readonly vehicles: Vehicle[];
  readonly depot: Location | null;
  readonly visits: Location[];
  readonly routes: RouteWithTrack[];
}

export enum ActionType {
  UPDATE_ROUTING_PLAN = 'UPDATE_ROUTING_PLAN',
  DELETE_LOCATION = 'DELETE_LOCATION',
  ADD_LOCATION = 'ADD_LOCATION',
  ADD_VEHICLE = 'ADD_VEHICLE',
  DELETE_VEHICLE = 'DELETE_VEHICLE',
  CLEAR_SOLUTION = 'CLEAR_SOLUTION',
}

export interface AddLocationAction extends Action<ActionType.ADD_LOCATION> {
  readonly value: LatLngWithDescription;
}

export interface AddVehicleAction extends Action<ActionType.ADD_VEHICLE> {
}

export interface ClearRouteAction extends Action<ActionType.CLEAR_SOLUTION> {
}

export interface DeleteLocationAction extends Action<ActionType.DELETE_LOCATION> {
  readonly value: number;
}

export interface DeleteVehicleAction extends Action<ActionType.DELETE_VEHICLE> {
  readonly value: number;
}

export interface VehicleCapacity {
  vehicleId: number;
  capacity: number;
}

export interface UpdateRouteAction extends Action<ActionType.UPDATE_ROUTING_PLAN> {
  readonly plan: RoutingPlan;
}

export type RouteAction =
  | AddLocationAction
  | AddVehicleAction
  | DeleteLocationAction
  | DeleteVehicleAction
  | UpdateRouteAction
  | ClearRouteAction;
