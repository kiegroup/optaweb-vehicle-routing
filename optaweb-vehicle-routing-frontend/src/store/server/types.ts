import { Action } from 'redux';
import { LatLng } from '../route/types';

export enum ActionType {
  SERVER_INFO = 'SERVER_INFO',
}

export interface ServerInfoAction extends Action<ActionType.SERVER_INFO> {
  value: ServerInfo;
}

export interface Demo {
  name: string;
  visits: number;
}

export type BoundingBox = [LatLng, LatLng];

export interface ServerInfo {
  boundingBox: BoundingBox | null;
  countryCodes: string[];
  demos: Demo[];
}
