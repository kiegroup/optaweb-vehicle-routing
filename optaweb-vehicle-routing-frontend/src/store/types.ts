/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Action } from 'redux';
import { ThunkAction } from 'redux-thunk';
import WebSocketClient from 'websocket/WebSocketClient';
import { UserViewport } from './client/types';
import { Demo } from './demo/types';
import { RoutingPlan } from './route/types';
import { ServerInfo } from './server/types';
import { WebSocketConnectionStatus } from './websocket/types';

/**
 * ThunkCommand is a ThunkAction that has no result (it's typically something like
 * `Promise<ActionAfterDataFetched>`, but sending messages over WebSocket usually has no response
 * (with the exception of subscribe), so most of our operations are void).
 *
 * @template A Type of action(s) allowed to be dispatched.
 */
export type ThunkCommand<A extends Action> = ThunkAction<void, AppState, WebSocketClient, A>;

/**
 * Factory method that takes a value and creates an @type {Action}.
 *
 * @template V value type
 * @template A action type
 */
export type ActionFactory<V, A extends Action> = V extends void ?
  // https://stackoverflow.com/questions/55646272/conditional-method-parameters-based-on-generic-type
  () => A : // nullary
  (value: V) => A; // unary

/**
 * Factory method that takes a value and creates a @type {ThunkCommand}.
 *
 * @template V value type
 * @template A action type
 */
export type ThunkCommandFactory<V, A extends Action> = V extends void ?
  () => ThunkCommand<A> : // nullary
  (value: V) => ThunkCommand<A>; // unary

export interface AppState {
  readonly serverInfo: ServerInfo;
  readonly plan: RoutingPlan;
  readonly connectionStatus: WebSocketConnectionStatus;
  readonly demo: Demo;
  readonly userViewport: UserViewport;
}
