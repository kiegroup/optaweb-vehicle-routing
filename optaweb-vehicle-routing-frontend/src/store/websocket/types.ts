import { Action } from 'redux';

export enum WebSocketConnectionStatus {
  OPEN = 'OPEN',
  CLOSED = 'CLOSED',
  ERROR = 'ERROR',
}

export enum ActionType {
  WS_CONNECT = 'WS_CONNECT',
  WS_CONNECT_SUCCESS = 'WS_CONNECT_SUCCESS',
  WS_CONNECT_FAILURE = 'WS_CONNECT_FAILURE',
}

export interface InitWsConnectionAction extends Action<ActionType.WS_CONNECT> {
}

export interface WsConnectionSuccessAction extends Action<ActionType.WS_CONNECT_SUCCESS> {
}

export interface WsConnectionFailureAction extends Action<ActionType.WS_CONNECT_FAILURE> {
  readonly error: string;
}

export type WebSocketAction =
  | InitWsConnectionAction
  | WsConnectionFailureAction
  | WsConnectionSuccessAction;
