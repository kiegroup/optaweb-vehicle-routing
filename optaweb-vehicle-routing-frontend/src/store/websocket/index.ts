import * as websocketOperations from './operations';
import { wsReducer } from './reducers';

export { websocketOperations };

export default wsReducer;
