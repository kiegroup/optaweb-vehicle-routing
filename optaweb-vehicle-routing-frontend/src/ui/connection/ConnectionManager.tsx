/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { UnpluggedIcon } from '@patternfly/react-icons';
import * as React from 'react';
import { connect } from 'react-redux';
import { AppState } from 'store/types';
import { websocketOperations } from 'store/websocket';
import { WebSocketConnectionStatus } from 'store/websocket/types';
import ConnectionError from 'ui/connection/ConnectionError';

interface StateProps {
  connectionStatus: WebSocketConnectionStatus;
}

interface DispatchProps {
  connectClient: typeof websocketOperations.connectClient;
}

export type Props = StateProps & DispatchProps;

const mapStateToProps = ({ connectionStatus }: AppState): StateProps => ({
  connectionStatus,
});

const mapDispatchToProps: DispatchProps = {
  connectClient: websocketOperations.connectClient,
};

export class ConnectionManager extends React.Component<Props> {
  componentDidMount() {
    this.props.connectClient();
  }

  render() {
    return (
      <ConnectionError
        isOpen={this.props.connectionStatus === WebSocketConnectionStatus.ERROR}
        message="Please check your network connection."
        title="Oops... Connection error!"
        icon={<UnpluggedIcon />}
        help="When connection is available the application will be functional again."
      />
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ConnectionManager);
