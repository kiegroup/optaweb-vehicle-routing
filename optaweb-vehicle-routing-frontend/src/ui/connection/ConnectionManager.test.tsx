/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { WebSocketConnectionStatus } from 'store/websocket/types';
import { ConnectionManager, Props } from './ConnectionManager';

describe('Connection Manager', () => {
  it('should connect the WebSocket client when mounted', () => {
    const props: Props = {
      connectClient: jest.fn(),
      connectionStatus: WebSocketConnectionStatus.CLOSED,
    };

    const connectionManager = shallow(<ConnectionManager {...props} />);
    expect(toJson(connectionManager)).toMatchSnapshot();

    // Connect WebSocket client when the component is mounted
    expect(props.connectClient).toHaveBeenCalled();
  });

  it('should not display error when connection is open', () => {
    const props: Props = {
      connectClient: jest.fn(),
      connectionStatus: WebSocketConnectionStatus.OPEN,
    };

    const connectionManager = shallow(<ConnectionManager {...props} />);
    expect(toJson(connectionManager)).toMatchSnapshot();
  });

  it('should display error when connection fails', () => {
    const props: Props = {
      connectClient: jest.fn(),
      connectionStatus: WebSocketConnectionStatus.ERROR,
    };

    const connectionManager = shallow(<ConnectionManager {...props} />);
    expect(toJson(connectionManager)).toMatchSnapshot();
  });
});
