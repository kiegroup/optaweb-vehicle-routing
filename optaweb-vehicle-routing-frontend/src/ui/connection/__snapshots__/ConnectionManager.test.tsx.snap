// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connection Manager should connect the WebSocket client when mounted 1`] = `
<ConnectionError
  help="When connection is available the application will be functional again."
  icon={
    <UnpluggedIcon
      color="currentColor"
      size="sm"
      title={null}
    />
  }
  isOpen={false}
  message="Please check your network connection."
  title="Oops... Connection error!"
/>
`;

exports[`Connection Manager should display error when connection fails 1`] = `
<ConnectionError
  help="When connection is available the application will be functional again."
  icon={
    <UnpluggedIcon
      color="currentColor"
      size="sm"
      title={null}
    />
  }
  isOpen={true}
  message="Please check your network connection."
  title="Oops... Connection error!"
/>
`;

exports[`Connection Manager should not display error when connection is open 1`] = `
<ConnectionError
  help="When connection is available the application will be functional again."
  icon={
    <UnpluggedIcon
      color="currentColor"
      size="sm"
      title={null}
    />
  }
  isOpen={false}
  message="Please check your network connection."
  title="Oops... Connection error!"
/>
`;
